zio-http/src/main/scala/zhttp/service/Handler.scala

```diff
@@ -119,29 +120,70 @@ private[zhttp] final case class Handler[R](
               else {
                 for {
-                  _ <- UIO { unsafeWriteAnyResponse(res) }
                   _ <- res.data match {
                     case HttpData.Empty =>
                       UIO {
+                        val size = res.data.unsafeSize
+                        unsafeWriteAnyResponse(
+                          res.updateHeaders(
+                            _ ++
+                              Headers(Name.ContentLength -> size.toString).when(size >= 0)
+                              ++ Headers(Name.TransferEncoding -> Value.Chunked).when(res.data.isChunked),
+                          ),
+                        )
                         unsafeWriteAndFlushLastEmptyContent()
                       }
 
                     case data @ HttpData.Text(_, _) =>
                       UIO {
-                        unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
+                        val byteBuf = data.encodeAndCache(res.attribute.memoize)
+                        val size    = byteBuf.readableBytes().toLong
+                        unsafeWriteAnyResponse(
+                          res.updateHeaders(
+                            _ ++
+                              Headers(Name.ContentLength -> size.toString).when(size >= 0)
+                              ++ Headers(Name.TransferEncoding -> Value.Chunked).when(res.data.isChunked),
+                          ),
+                        )
+                        unsafeWriteLastContent(byteBuf)
                       }
 
                     case HttpData.BinaryByteBuf(data) =>
                       UIO {
+                        val size = res.data.unsafeSize
+                        unsafeWriteAnyResponse(
+                          res.updateHeaders(
+                            _ ++
+                              Headers(Name.ContentLength -> size.toString).when(size >= 0)
+                              ++ Headers(Name.TransferEncoding -> Value.Chunked).when(res.data.isChunked),
+                          ),
+                        )
                         unsafeWriteLastContent(data)
                       }
 
                     case data @ HttpData.BinaryChunk(_) =>
                       UIO {
-                        unsafeWriteLastContent(data.encodeAndCache(res.attribute.memoize))
+                        val byteBuf = data.encodeAndCache(res.attribute.memoize)
+                        val size    = byteBuf.readableBytes().toLong
+                        unsafeWriteAnyResponse(
+                          res.updateHeaders(
+                            _ ++
+                              Headers(Name.ContentLength -> size.toString).when(size >= 0)
+                              ++ Headers(Name.TransferEncoding -> Value.Chunked).when(res.data.isChunked),
+                          ),
+                        )
+                        unsafeWriteLastContent(byteBuf)
```

Comment:
Only for Text and BinaryChunk headers are computed via ByteBuf returned by encodeAndCache